src/data/data.cu

```diff
@@ -12,75 +12,86 @@
 
 namespace xgboost {
 
+template <typename T>
+void CopyInfoImpl(std::map<std::string, Json> const& column, HostDeviceVector<float>* out) {
+  auto SetDeviceToPtr = [](void* ptr) {
+    cudaPointerAttributes attr;
+    dh::safe_cuda(cudaPointerGetAttributes(&attr, ptr));
+    int32_t ptr_device = attr.device;
+    dh::safe_cuda(cudaSetDevice(ptr_device));
+    return ptr_device;
+  };
+
+  common::Span<T> s_data { ArrayInterfaceHandler::ExtractData<T>(column) };
+  auto ptr_device = SetDeviceToPtr(s_data.data());
+  thrust::device_ptr<T> p_src {s_data.data()};
+
+  auto length = s_data.size();
+  out->SetDevice(ptr_device);
+  out->Resize(length);
+
+  auto p_dst = thrust::device_pointer_cast(out->DevicePointer());
+  thrust::copy(p_src, p_src + length, p_dst);
```

Comment:
Just a note that thrust copy generally should be avoided. This is because if the data is non-trivially copyable (defines a copy constructor) thrust must stage the copy in temporary buffers in both host and device memory.

In this case I think it's fine because the type is trivial, but I tend to avoid this function completely because of large hidden performance penalties that are hard to understand for newer developers. You can end up in a situation where someone adds a constructor to a trivial type and suddenly memory usage is much larger and performance is worse.

Reply:
@RAMitchell Thanks for the explanation.  I will use plain memcpy.